app/backend/approaches/chatreadretrieveread.py

```diff
@@ -117,3 +114,66 @@ def get_chat_history_as_text(self, history: Sequence[dict[str, str]], include_la
             if len(history_text) > approx_max_tokens*4:
                 break    
         return history_text
+    
+    # Generate messages needed for chat Completion api
+    def get_messages_from_history(self, prompt_override, follow_up_questions_prompt, history: Sequence[dict[str, str]], sources: str, approx_max_tokens: int = 1000) -> []:
+        messages = []
+        token_count = 0
+        if prompt_override is None:
+            system_message = self.system_message_chat_conversation.format(injected_prompt="", follow_up_questions_prompt=follow_up_questions_prompt)
+        elif prompt_override.startswith(">>>"):
+            system_message = self.system_message_chat_conversation.format(injected_prompt=prompt_override[3:] + "\n", follow_up_questions_prompt=follow_up_questions_prompt)
+        else:
+            system_message = prompt_override.format(follow_up_questions_prompt=follow_up_questions_prompt)
+
+        messages.append({"role":self.SYSTEM, "content": system_message})
+        token_count += self.num_tokens_from_messages(messages[-1], self.chatgpt_model)
+        
+        #latest conversation
+        user_content = history[-1]["user"] + " \nSources:" + sources
+        messages.append({"role": self.USER, "content": user_content})
+        token_count += token_count + self.num_tokens_from_messages(messages[-1], self.chatgpt_model)
+
+        '''
+        Enqueue in reverse order
+        if limit exceeds truncate old messages 
+        leaving system message behind
+        Keep track of token count for each conversation
+        If token count exceeds limit, break
+        '''
+        for h in reversed(history[:-1]):
+            if h.get("bot"):
+                messages.insert(1, {"role": self.ASSISTANT, "content" : h.get("bot")})
+                token_count += self.num_tokens_from_messages(messages[1], self.chatgpt_model)
+            messages.insert(1, {"role": self.USER, "content" : h.get("user")})
+            token_count += self.num_tokens_from_messages(messages[1], self.chatgpt_model)
+            if token_count > approx_max_tokens*4:
+                break
+        return messages
+    
+    def num_tokens_from_messages(self, message: dict[str,str], model: str):
+        """
+        Calculate the number of tokens required to encode a message.
+        Args:
+            message (any): The message to encode, represented as a dictionary.
+            model (str): The name of the model to use for encoding.
+        Returns:
+            int: The total number of tokens required to encode the message.
+        Example:
+            message = {'role': 'user', 'name': 'John', 'content': 'Hello, how are you?'}
+            model = 'gpt-3.5-turbo'
+            num_tokens_from_messages(message, model)
+            output: 11
+        """
+        encoding = tiktoken.encoding_for_model(self.get_oai_chatmodel_tiktok(model))
+        num_tokens = 0
+        num_tokens += 2  # every message follows {role/name}\n{content}\n
```

Comment:
The 2 is for the keys, right?

If so, this inline comment seems clearer?

```suggestion
        num_tokens += 2  # For "role" and "content" keys
```

If its not for the keys, then I'm confused.